test/unittests/instantiate_test.cpp

```diff
@@ -24,6 +24,75 @@ TEST(instantiate, imported_functions)
     EXPECT_EQ(instance.imported_function_types[0], TypeIdx{0});
 }
 
+TEST(instantiate, imported_globals)
+{
+    Module module;
+    module.importsec.emplace_back(Import{"mod", "g", ExternalKind::Global, {true}});
+
+    uint64_t global_value = 42;
+    ImportedGlobal g{&global_value, true};
+    auto instance = instantiate(module, {}, {g});
+
+    ASSERT_EQ(instance.imported_globals.size(), 1);
+    EXPECT_EQ(instance.imported_globals[0].is_mutable, true);
+    EXPECT_EQ(*instance.imported_globals[0].value, 42);
+    ASSERT_EQ(instance.globals.size(), 0);
+}
+
+TEST(instantiate, imported_globals_multiple)
+{
+    Module module;
+    module.importsec.emplace_back(Import{"mod", "g1", ExternalKind::Global, {true}});
+    module.importsec.emplace_back(Import{"mod", "g2", ExternalKind::Global, {false}});
+
+    uint64_t global_value1 = 42;
+    ImportedGlobal g1{&global_value1, true};
+    uint64_t global_value2 = 43;
+    ImportedGlobal g2{&global_value2, false};
+    auto instance = instantiate(module, {}, {g1, g2});
+
+    ASSERT_EQ(instance.imported_globals.size(), 2);
+    EXPECT_EQ(instance.imported_globals[0].is_mutable, true);
+    EXPECT_EQ(instance.imported_globals[1].is_mutable, false);
+    EXPECT_EQ(*instance.imported_globals[0].value, 42);
+    EXPECT_EQ(*instance.imported_globals[1].value, 43);
+    ASSERT_EQ(instance.globals.size(), 0);
+}
+
+TEST(instantiate, imported_globals_mismatched_count)
+{
+    Module module;
+    module.importsec.emplace_back(Import{"mod", "g1", ExternalKind::Global, {true}});
+    module.importsec.emplace_back(Import{"mod", "g2", ExternalKind::Global, {false}});
+
+    uint64_t global_value = 42;
+    ImportedGlobal g{&global_value, true};
+    EXPECT_THROW(instantiate(module, {}, {g}), std::runtime_error);
+}
+
+TEST(instantiate, imported_globals_mismatched_mutability)
+{
+    Module module;
+    module.importsec.emplace_back(Import{"mod", "g1", ExternalKind::Global, {true}});
+    module.importsec.emplace_back(Import{"mod", "g2", ExternalKind::Global, {false}});
+
+    uint64_t global_value1 = 42;
+    ImportedGlobal g1{&global_value1, false};
+    uint64_t global_value2 = 42;
+    ImportedGlobal g2{&global_value2, true};
+    EXPECT_THROW(instantiate(module, {}, {g1, g2}), std::runtime_error);
+}
+
+TEST(instantiate, imported_globals_nullptr)
+{
+    Module module;
+    module.importsec.emplace_back(Import{"mod", "g1", ExternalKind::Global, {true}});
+    module.importsec.emplace_back(Import{"mod", "g2", ExternalKind::Global, {false}});
+
+    ImportedGlobal g{nullptr, true};
+    EXPECT_THROW(instantiate(module, {}, {g}), std::runtime_error);
```

Comment:
I believe this fails because of the number of globals mismatch (https://github.com/wasmx/fizzy/pull/77/files#diff-da8232128fc88a65268ab30bf8e8abdcR33), not because of the `nullptr`.

Unfortunately, GTest do not provide a macro for testing exception message, but it might be good time to add one ourselves. The https://github.com/wasmx/fizzy/pull/77/files#diff-da8232128fc88a65268ab30bf8e8abdcR33 provides some possible implementations of this.

Reply:
Yeah, good catch. In this PR I would just pass the right number of globals here.

(you mistyped the second link)

Also not sure if it makes sense to test the exception messages at all, but if/when we add the error codes, it certainly makes sense to test those.

Reply:
https://github.com/wasmx/fizzy/pull/94